### ISSUE TITLE
example for is next sentence

### ISSUE THREAD

**Issue by charlesmartin14** [huggingface/pytorch-pretrained-BERT#48]
Can you make up a working example for 'is next sentence'

Is this expected to work properly ?

```
# Load pre-trained model tokenizer (vocabulary)
tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')

# Tokenized input
text = "Who was Jim Morrison ? Jim Morrison was a puppeteer"
tokenized_text = tokenizer.tokenize(text)

# Convert token to vocabulary indices
indexed_tokens = tokenizer.convert_tokens_to_ids(tokenized_text)
# Define sentence A and B indices associated to 1st and 2nd sentences (see paper)
segments_ids = [0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1]

# Convert inputs to PyTorch tensors
tokens_tensor = torch.tensor([indexed_tokens])
segments_tensors = torch.tensor([segments_ids])

# Load pre-trained model (weights)
model = BertForNextSentencePrediction.from_pretrained('bert-base-uncased')
model.eval()

# Predict is Next Sentence ?
predictions = model(tokens_tensor, segments_tensors)
```

**Comment by thomwolf** [comment]
I think it should work. You should get a [1, 2] tensor of logits where `predictions[0, 0]` is the score of Next sentence being `True` and `predictions[0, 1]` is the score of Next sentence being `False`. So just take the max of the two (or use a `SoftMax` to get probabilities).
Did you try it?
The model behaves better on longer sentences of course (it's mainly trained on 512 tokens inputs).

**Comment by thomwolf** [comment]
Closing that for now, feel free to reopen if there is another issue.

**Comment by charlesmartin14** [comment]
Thank you
 I should have clarified...yes this is exactly what I did, but the answer was surprisingly bad

For example

`text = "Who was Jim Morrison ? Apples taste better than oranges "
...
predictions
`
gives
`
tensor([[ 1.6175, -0.4209]], grad_fn=<ThAddmmBackward>)
`

I suppose the better thing to do is test a random sample of the test sentences
thanks again

**Comment by Alexadar** [comment]
Guys, are [CLS] and [SEP] tokens mandatory for this example?

**Comment by dirkgr** [comment]
This is not super clear, even wrong in the examples, but there is this note in the docstring for `BertModel`:
```
        `pooled_output`: a torch.FloatTensor of size [batch_size, hidden_size] which is the output of a
            classifier pretrained on top of the hidden state associated to the first character of the
            input (`CLF`) to train on the Next-Sentence task (see BERT's paper).
```
That seems to suggest pretty strongly that you have to put in the `CLF` token.

**Comment by hackable** [comment]
```import torch
from pytorch_pretrained_bert import BertTokenizer, BertModel, BertForMaskedLM,BertForNextSentencePrediction

# Load pre-trained model tokenizer (vocabulary)
tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')

# Tokenized input
text = "[CLS] Who was Jim Henson ? [SEP] Jim Henson was a puppeteer [SEP]"
tokenized_text = tokenizer.tokenize(text)

# Convert token to vocabulary indices
indexed_tokens = tokenizer.convert_tokens_to_ids(tokenized_text)
# Define sentence A and B indices associated to 1st and 2nd sentences (see paper)
segments_ids = [0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1]

# Convert inputs to PyTorch tensors
tokens_tensor = torch.tensor([indexed_tokens])
segments_tensors = torch.tensor([segments_ids])

# Load pre-trained model (weights)
model = BertForNextSentencePrediction.from_pretrained('bert-base-uncased')
model.eval()

# Predict is Next Sentence ?
predictions = model(tokens_tensor, segments_tensors, attention_mask=None, torch.tensor([0,1]) )




print(predictions)
`

Result I get is tensor([[ 6.3714, -6.3910]], grad_fn=<AddmmBackward>)

How do i infer this as true or false

**Comment by dariocazzani** [comment]
Those are the logits, because you did not pass the `next_sentence_label`.

My understanding is that you could apply a softmax and get the probability for the sequence to be a possible sequence.

`Sentence 1: How old are you?`
`Sentence 2: The Eiffel Tower is in Paris`
`tensor([[-2.3808,  5.4018]], grad_fn=<AddmmBackward>)`
`Sentence 1: How old are you?`
`Sentence 2: I am 193 years old`
`tensor([[ 6.0164, -5.7138]], grad_fn=<AddmmBackward>)`

For the first example the probability that the second sentence is a probable continuation is very low.
For the second example the probability is very high (I am looking at the first logit)

**Comment by pengjiao123** [comment]
predictions = model(tokens_tensor, segments_tensors )
I try the code more than once，why I have the different result？
sometime predictions[0, 0] is higher ，however， the same sentence pair，predictions[0, 0] is lower.

**Comment by thomwolf** [comment]
Maybe your model is not in evaluation mode (`model.eval()`)?
You need to do this to desactivate the dropout modules.

**Comment by pengjiao123** [comment]
It is OK.THANKS A LOT.

**Comment by AIGyan** [comment]
error: 
--> 197         embeddings = words_embeddings + position_embeddings + token_type_embeddings
    198         embeddings = self.LayerNorm(embeddings)
    199         embeddings = self.dropout(embeddings)
The size of tensor a (21) must match the size of tensor b (14) at non-singleton dimension 1

The above issues get resolved, when I added few extra 1's and 0's to make the shape similar tokens_tensor and segments_tensors. Just wondering am I using in a right way. 

My predictions output is a tensor array of size 21 X 30522 . 
And what I believe the example is to predict the word which is [MASK] . Can you also please guide how to predict the next sentence?

**Comment by yuchuang1979** [comment]
> Maybe your model is not in evaluation mode (`model.eval()`)?
> You need to do this to desactivate the dropout modules.

@thomwolf Actually even when I used model.eval() I still got different results. I observed this when I use every model of the package (BertModel, BertForNextSentencePrediction etc). Only when I fixed the length of the input (e.g. to 128), I can get the same results. In this way I have to pad 0 to indexed_tokens so it has a fixed length.

Could you explain why is like this, or did I make any mistake?

Thank you so much!

**Comment by Alexadar** [comment]
> > Maybe your model is not in evaluation mode (`model.eval()`)?
> > You need to do this to desactivate the dropout modules.
> 
> @thomwolf Actually even when I used model.eval() I still got different results. I observed this when I use every model of the package (BertModel, BertForNextSentencePrediction etc). Only when I fixed the length of the input (e.g. to 128), I can get the same results. In this way I have to pad 0 to indexed_tokens so it has a fixed length.
> 
> Could you explain why is like this, or did I make any mistake?
> 
> Thank you so much!

Make sure
1) input_ids, input_mask, segment_ids have same length
2) vocabulary file for tokenizer is from the same config as your bert_config.json

I have symilar symptoms when vocab and config was diferent

**Comment by yuchuang1979** [comment]
> > @thomwolf Actually even when I used model.eval() I still got different results. I observed this when I use every model of the package (BertModel, BertForNextSentencePrediction etc). Only when I fixed the length of the input (e.g. to 128), I can get the same results. In this way I have to pad 0 to indexed_tokens so it has a fixed length.
> > Could you explain why is like this, or did I make any mistake?
> > Thank you so much!
> 
> Make sure
> 
> 1. input_ids, input_mask, segment_ids have same length
> 2. vocabulary file for tokenizer is from the same config dir as your bert_config.json
> 
> I had symilar symptoms when vocab and config was from diferent berts

@Alexadar thanks for your reply! I did everything as you mentioned. Now I noticed that after I load the model {e.g., config = ppb.BertConfig.from_pretrained(init_checkpoint_pt);
model = ppb.BertModel(config)}, I can get the same results if I execute 'outputs = model(tokens_tensor)' for several times.

However, if I load the model again and execute  'outputs = model(tokens_tensor)' on the same sentence, I will get a new result. 

This may cause problems in some applications - each time I will get a new result (e.g. encoding of a sentence) for the same sentence.

**Comment by pbabvey** [comment]
I noticed that the probability for longer sentences, regardless of how much they are related to the same subject, is higher than the shorter ones. For example, I added some random sentences to the end of the first or second part and observed significant increase in the first logit value. Is it a way to regularize the model for the next sentence prediction?

**Comment by ehsan-soe** [comment]
@pbabvey I observe increase in the second logit though.

**Comment by ehsan-soe** [comment]
@pbabvey I am observing the same thing.
Is the probabilities length normalized?

**Comment by ajbot2019** [comment]
import torch 
from transformers import BertTokenizer, BertModel, BertForMaskedLM,BertForNextSentencePrediction
tokenizer=BertTokenizer.from_pretrained('bert-base-uncased')
BertNSP=BertForNextSentencePrediction.from_pretrained('bert-base-uncased')


text1 = "How old are you?"
text2 = "The Eiffel Tower is in Paris"

text1_toks = ["[CLS]"] + tokenizer.tokenize(text1) + ["[SEP]"]
text2_toks = tokenizer.tokenize(text2) + ["[SEP]"]
text=text1_toks+text2_toks
print(text)
indexed_tokens = tokenizer.convert_tokens_to_ids(text1_toks + text2_toks)
segments_ids = [0]*len(text1_toks) + [1]*len(text2_toks)

tokens_tensor = torch.tensor([indexed_tokens])
segments_tensors = torch.tensor([segments_ids])
print(indexed_tokens)
print(segments_ids)
BertNSP.eval()
prediction = BertNSP(tokens_tensor, segments_tensors)
prediction=prediction[0]  # tuple to tensor
print(predictions)
softmax = torch.nn.Softmax(dim=1)
prediction_sm = softmax(prediction)
print (prediction_sm)

o/p of predictions
tensor([[ 2.1772, -0.8097]], grad_fn=<AddmmBackward>)

o/p of prediction_sm
tensor([[0.9923, 0.0077]], grad_fn=<SoftmaxBackward>)


why is the score still high 0.9923 even after apply softmax ?

**Comment by ajbot2019** [comment]
> Those are the logits, because you did not pass the `next_sentence_label`.
> 
> My understanding is that you could apply a softmax and get the probability for the sequence to be a possible sequence.
> 
> `Sentence 1: How old are you?`
> `Sentence 2: The Eiffel Tower is in Paris`
> `tensor([[-2.3808, 5.4018]], grad_fn=<AddmmBackward>)`
> `Sentence 1: How old are you?`
> `Sentence 2: I am 193 years old`
> `tensor([[ 6.0164, -5.7138]], grad_fn=<AddmmBackward>)`
> 
> For the first example the probability that the second sentence is a probable continuation is very low.
> For the second example the probability is very high (I am looking at the first logit)

im getting different scores for the sentences that you have tried  . please advise why i'm getting it below is my code .

import torch
from transformers import BertTokenizer, BertModel, BertForMaskedLM,BertForNextSentencePrediction
tokenizer=BertTokenizer.from_pretrained('bert-base-uncased')
BertNSP=BertForNextSentencePrediction.from_pretrained('bert-base-uncased')

text1 = "How old are you?"
text2 = "The Eiffel Tower is in Paris"

text1_toks = ["[CLS]"] + tokenizer.tokenize(text1) + ["[SEP]"]
text2_toks = tokenizer.tokenize(text2) + ["[SEP]"]
text=text1_toks+text2_toks
print(text)
indexed_tokens = tokenizer.convert_tokens_to_ids(text1_toks + text2_toks)
segments_ids = [0]*len(text1_toks) + [1]*len(text2_toks)

tokens_tensor = torch.tensor([indexed_tokens])
segments_tensors = torch.tensor([segments_ids])
print(indexed_tokens)
print(segments_ids)
BertNSP.eval()
prediction = BertNSP(tokens_tensor, segments_tensors)
prediction=prediction[0] # tuple to tensor
print(predictions)
softmax = torch.nn.Softmax(dim=1)
prediction_sm = softmax(prediction)
print (prediction_sm)

o/p of predictions
tensor([[ 2.1772, -0.8097]], grad_fn=)

o/p of prediction_sm
tensor([[0.9923, 0.0077]], grad_fn=)

why is the score still high 0.9923 even after apply softmax ?

**Comment by parth126** [comment]
> > Those are the logits, because you did not pass the `next_sentence_label`.
> > My understanding is that you could apply a softmax and get the probability for the sequence to be a possible sequence.
> > `Sentence 1: How old are you?`
> > `Sentence 2: The Eiffel Tower is in Paris`
> > `tensor([[-2.3808, 5.4018]], grad_fn=<AddmmBackward>)`
> > `Sentence 1: How old are you?`
> > `Sentence 2: I am 193 years old`
> > `tensor([[ 6.0164, -5.7138]], grad_fn=<AddmmBackward>)`
> > For the first example the probability that the second sentence is a probable continuation is very low.
> > For the second example the probability is very high (I am looking at the first logit)
> 
> im getting different scores for the sentences that you have tried . please advise why i'm getting it below is my code .
> 
> import torch
> from transformers import BertTokenizer, BertModel, BertForMaskedLM,BertForNextSentencePrediction
> tokenizer=BertTokenizer.from_pretrained('bert-base-uncased')
> BertNSP=BertForNextSentencePrediction.from_pretrained('bert-base-uncased')
> 
> text1 = "How old are you?"
> text2 = "The Eiffel Tower is in Paris"
> 
> text1_toks = ["[CLS]"] + tokenizer.tokenize(text1) + ["[SEP]"]
> text2_toks = tokenizer.tokenize(text2) + ["[SEP]"]
> text=text1_toks+text2_toks
> print(text)
> indexed_tokens = tokenizer.convert_tokens_to_ids(text1_toks + text2_toks)
> segments_ids = [0]*len(text1_toks) + [1]*len(text2_toks)
> 
> tokens_tensor = torch.tensor([indexed_tokens])
> segments_tensors = torch.tensor([segments_ids])
> print(indexed_tokens)
> print(segments_ids)
> BertNSP.eval()
> prediction = BertNSP(tokens_tensor, segments_tensors)
> prediction=prediction[0] # tuple to tensor
> print(predictions)
> softmax = torch.nn.Softmax(dim=1)
> prediction_sm = softmax(prediction)
> print (prediction_sm)
> 
> o/p of predictions
> tensor([[ 2.1772, -0.8097]], grad_fn=)
> 
> o/p of prediction_sm
> tensor([[0.9923, 0.0077]], grad_fn=)
> 
> why is the score still high 0.9923 even after apply softmax ?

I am facing the same issue. No matter what sentences I use, I always get very high probability of the second sentence being related to the first.

**Comment by LysandreJik** [comment]
@parth126 have you seen https://github.com/huggingface/transformers/issues/1788 and is it related to your issue?

**Comment by parth126** [comment]
> @parth126 have you seen #1788 and is it related to your issue?

Yes it was the same issue. And the solution worked like a charm. 
Many thanks @LysandreJik
